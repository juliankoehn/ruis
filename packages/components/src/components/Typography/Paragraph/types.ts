export type ParagraphTypes = 'p' | 'span' | 'div'

export const defaultProps = {
    is: 'p',
    testId: 'ruids-paragraph',
}

export type ParagraphProps = {
    is: ParagraphTypes
    testId?: string
    style?: React.CSSProperties
    className?: string
    children?: React.ReactNode
} & typeof defaultProps