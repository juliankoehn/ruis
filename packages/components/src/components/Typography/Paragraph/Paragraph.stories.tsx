import React from 'react'
import { withKnobs } from '@storybook/addon-knobs'
import { ParagraphTypes } from './types'
import { Paragraph } from './Paragraph'

export default {
    title: 'Typography/Paragraph',
    decorators: [withKnobs]
}

const paragraphTypes: ParagraphTypes[] = ['p', 'div', 'span']

export const Default = () => {
    return (
        <div>
            {paragraphTypes.map(size => (
                <Paragraph is={size}>Heading {size}</Paragraph>
            ))}
        </div>

    )
}
