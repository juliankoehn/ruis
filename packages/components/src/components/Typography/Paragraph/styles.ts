import { css } from '@emotion/core'
import tokens from '@ruids/tokens'
import { ParagraphTypes } from './types';

export const getParagraphStyles = (spacing: string, element: ParagraphTypes) => {
    return css`
        display: ${element === 'p' || element === 'div' ? 'block' : undefined};
        font-family: ${tokens.fontSans};
        color: ${tokens.colorTextDark};
        margin-top: 0;
        margin-bottom: ${tokens[spacing]};
        font-weight: ${tokens.fontNormal};
        font-size: ${tokens.textSm};
        line-height: ${tokens.leadingNormal};
    `
};