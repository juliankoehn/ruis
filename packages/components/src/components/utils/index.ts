export { canUseDOM } from './canUseDom'
export { lightenColor } from './color.lighten'
export {Â axe } from './axe'