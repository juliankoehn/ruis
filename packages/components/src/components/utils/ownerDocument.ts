export const ownerDocument = (node: Node): Document => {
  return (node && node.ownerDocument) || document;
};
