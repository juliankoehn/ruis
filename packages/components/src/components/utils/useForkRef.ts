import { useMemo, MutableRefObject } from 'react';

export type HTMLElementOrNull = HTMLElement | null;
export type CallbackRef = (node: HTMLElement | null) => any;
export type AnyRef = CallbackRef | MutableRefObject<HTMLElementOrNull>;

function setRef(ref: AnyRef, value: HTMLElementOrNull) {
  if (typeof ref === 'function') {
    ref(value);
  } else if (ref) {
    ref.current = value;
  }
}

function useForkRef(refA: AnyRef, refB: AnyRef): CallbackRef | null {
  /**
   * This will create a new function if the ref props change and are defined.
   * This means react will call the old forkRef with `null` and the new forkRef
   * with the ref. Cleanup naturally emerges from this behavior
   */
  return useMemo(() => {
    if (refA == null && refB == null) {
      return null;
    }
    return (refValue: HTMLElementOrNull) => {
      setRef(refA, refValue);
      setRef(refB, refValue);
    };
  }, [refA, refB]);
}

export { useForkRef };
