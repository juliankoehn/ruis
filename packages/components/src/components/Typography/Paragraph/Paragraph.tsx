import React from 'react'
import { ParagraphProps, defaultProps } from './types'
import { TypographyContext } from '../Typography'
import { getParagraphStyles } from './styles'

export class Paragraph extends React.Component<ParagraphProps> {
    static defaultProps = defaultProps

    render() {
        const {
            className,
            children,
            is,
            testId,
            ...otherProps
        } = this.props
        const Element = is

        
        return (
            <TypographyContext.Consumer>
                {value => (
                    <Element
                        className={className}
                        data-test-id={testId}
                        {...otherProps}
                        css={getParagraphStyles(value.paragraph.spacing, is)}
                    >
                        {children}
                    </Element>
                )}
            </TypographyContext.Consumer>
        )
    }
}