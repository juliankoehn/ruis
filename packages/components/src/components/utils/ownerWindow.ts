import { ownerDocument } from './ownerDocument';

export const ownerWindow = (node: Node): Window => {
  const doc = ownerDocument(node);
  return doc.defaultView || window;
};
