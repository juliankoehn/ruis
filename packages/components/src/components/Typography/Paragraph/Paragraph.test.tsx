import React from 'react';
import Paragraph from './Paragraph';
import { render } from '@testing-library/react';
import { axe } from '../../utils';

it('renders the component', () => {
  const { container } = render(<Paragraph>Paragraph</Paragraph>);

  expect(container).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(
    <Paragraph className="my-extra-class">Paragraph</Paragraph>,
  );

  expect(container).toMatchSnapshot();
});

it('renders the component lg', () => {
  const { container } = render(<Paragraph size="lg">Paragraph</Paragraph>);

  expect(container).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(<Paragraph>Paragraph</Paragraph>);

  expect(await axe(container)).toHaveNoViolations();
});
