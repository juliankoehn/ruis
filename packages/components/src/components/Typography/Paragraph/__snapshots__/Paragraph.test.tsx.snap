// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div>
  <p
    class="css-g3ss1z"
    data-test-id="ruids-paragraph"
  >
    Paragraph
  </p>
</div>
`;

exports[`renders the component span 1`] = `
<div>
  <span
    class="css-1h5b7sa"
    data-test-id="ruids-paragraph"
  >
    Paragraph
  </span>
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div>
  <p
    class="my-extra-class css-g3ss1z"
    data-test-id="ruids-paragraph"
  >
    Paragraph
  </p>
</div>
`;
